galsim/config/extra.py

```diff
@@ -0,0 +1,377 @@
+                raise ValueError("Cannot skip hdus.  Not output found for hdu %d"%h)
+        # Turn hdus into a list (in order)
+        hdulist = [ hdus[k] for k in range(first,len(hdus)+first) ]
+        return hdulist
+    else:
+        return []
+
+valid_extra_outputs = {}
+
+def RegisterExtraOutput(key, init_func=None, kwargs_func=None, setup_func=None, stamp_func=None,
+                        image_func=None, final_func=None, write_func=None, hdu_func=None):
+    """Register an extra output field for use by the config apparatus.
+
+    @param key              The name of the output field in config['output']
+    @param init_func        A function or class name to use to build the output object. 
+                            [default: None, in which case the output "object" will be a list
+                            of length nimages that can be used to construct what is needed.]
+    @param kwargs_func      A function to get the initialization kwargs. [default: None, which
+                            means initialize with no arguments.]
+    @param setup_func       A function to call at the start of each image. 
+                            The call signature is
+                                Setup(output_obj, scratch, config, base, logger)
+    @param stamp_func       A function to call at the end of building each stamp.
+                            The call signature is 
+                                ProcessStamp(output_obj, scratch, config, base, obj_num, logger)
+    @param image_func       A function to call at the end of building each image
+                            The call signature is 
+                                ProcessImage(output_obj, scratch, config, base, obj_nums, logger)
+                            where obj_nums is a list of the object numbers used for this image.
+    @param final_func       A function to call at the end of the file processing to construct
+                            the final object to be written. [default: None, which means the
+                            init_func already created the correct object, so just use that.]
+                            The call signature is 
+                                output_obj = Finalize(output_obj, scratch, config, base, logger)
+    @param write_func       A function to call to write the output file
+                            The call signature is 
+                                WriteFile(output_obj, file_name)
+    @param hdu_func         A function to call to build a FITS HDU or an Image to put in an HDU.
+                            The call signature is   
+                                hdu = WriteToHDU(output_obj)
+    """
+    valid_extra_outputs[key] = {
+        'init' : init_func,
+        'kwargs' : kwargs_func,
+        'setup' : setup_func,
+        'stamp' : stamp_func,
+        'image' : image_func,
+        'final' : final_func,
+        'write' : write_func,
+        'hdu' : hdu_func
+    }
```

Comment:
OK.  I switched the extra output stuff to use the inheritance model, and I agree that it is a nicer approach, so I'm convinced.  

I had to be a bit careful about some of the multiprocessing communication, but I think I got it all sorted out.  The key is to give the builder the scratch and data objects to use for writing anything that you want to persist in later stages.

I'll work on adopting this model for the other registrations.  For now if you want to check it out, here is the commit for the extra output types: f058762dab0415073b2e57dd6881b672f10b4567